### PR TITLE
Pre-review xen ffa v7 patches

### PR DESCRIPTION
**Pull request by jenswikl**
This is a dummy pull request used to get a view of the v7 patchset in progress

### CODE REVIEW

**Review by bertrand-marquis — COMMENTED** [review]
This is something not compliant with the spec so you will need to explain why exactly this is required.

File: xen/arch/arm/vsmc.c
Comment (bertrand-marquis) [inline xen/arch/arm/vsmc.c:89]: Correct me if I am wrong but here you make Xen claim it is supporting 1.2 version but in practice all calls would still be using 1.1

File: xen/arch/arm/include/asm/smccc.h
Comment (bertrand-marquis) [inline xen/arch/arm/include/asm/smccc.h:281]: We should use a fixed size type here. I am wondering wether we should use uint64_t or register_t

**Comment by jenswikl** [comment]
> This is something not compliant with the spec so you will need to explain why exactly this is required.

Yeah, that's why I've asked for opinions on this. The kernel does this and breaks my setup without the "xen/arm: ffa: relaxed FFA_PARTITION_INFO_GET_COUNT_FLAG support" patch.

**Comment by bertrand-marquis** [comment]
so this should be mentioned on your commit message as it explains why you actually need to do this.
Also be prepared for the xen community to ask you to go and fix this in the kernel instead of trying to workaround the issue inside xen

File: xen/arch/arm/include/asm/smccc.h
Comment (jenswikl) [reply xen/arch/arm/include/asm/smccc.h:281]: This was reviewed at https://lore.kernel.org/all/C8575668-5771-4E77-B77C-15E6E740E540@arm.com/

I'll keep it as it is for now and let you comment on the mailing list.

File: xen/arch/arm/vsmc.c
Comment (jenswikl) [reply xen/arch/arm/vsmc.c:89]: This was reviewed at https://lore.kernel.org/all/0c6d58ae-bd8d-a3d8-d8d9-2b7a5340b4dd@xen.org/

**Comment by jenswikl** [comment]
It's not entirely clear to me how much more the commits should be split up in order to make them smaller.
These are fairly small, < 130 lines:
- xen/arm: ffa: relaxed FFA_PARTITION_INFO_GET_COUNT_FLAG support
- xen/arm: move regpair_to_uint64() and uint64_to_regpair() to regs.h
- xen/arm: ffa: support guest FFA_PARTITION_INFO_GET
- xen/arm: ffa: support mapping guest RX/TX buffers
- xen/arm: ffa: map SPMC rx/tx buffers
- xen/arm: ffa: add direct request support
- xen/arm: smccc: add support for SMCCCv1.2 extended input/output registers

This one is close to 200 lines, but I'm not sure how I should split it to make it easier to review. For instance, if I add unused static functions they are bound to cause compiler warnings.
- xen/arm: ffa: send guest events to Secure Partitions

This introduces the mediator and is quite large, but it's complete and only does the bare minimum. I could reduce the largest file, xen/arch/arm/ffa.c, quite a bit. Will that be enough? If this should be taken further how would you recommend splitting it? Tools changes and docs in one commit and the rest in another?
- xen/arm: add a primitive FF-A mediator

I'll try to take this one in steps, all structs in one commit, unfragmented sharing in one and finally fragmented sharing in the last.
- xen/arm: ffa: support sharing memory

**Comment by bertrand-marquis** [comment]
Hi Jens,
> It's not entirely clear to me how much more the commits should be split up in order to make them smaller. These are fairly small, < 130 lines:
> 
> * xen/arm: ffa: relaxed FFA_PARTITION_INFO_GET_COUNT_FLAG support
> * xen/arm: move regpair_to_uint64() and uint64_to_regpair() to regs.h
> * xen/arm: ffa: support guest FFA_PARTITION_INFO_GET
> * xen/arm: ffa: support mapping guest RX/TX buffers
> * xen/arm: ffa: map SPMC rx/tx buffers
> * xen/arm: ffa: add direct request support
> * xen/arm: smccc: add support for SMCCCv1.2 extended input/output registers
> 
> This one is close to 200 lines, but I'm not sure how I should split it to make it easier to review. For instance, if I add unused static functions they are bound to cause compiler warnings.
> 
> * xen/arm: ffa: send guest events to Secure Partitions

All those are ok i think in term of size.

> 
> This introduces the mediator and is quite large, but it's complete and only does the bare minimum. I could reduce the largest file, xen/arch/arm/ffa.c, quite a bit. Will that be enough? If this should be taken further how would you recommend splitting it? Tools changes and docs in one commit and the rest in another?

Definitely you must split the tool changes and put it in a separate patch for 2 reasons:
- it will reduce the size
- tools and docs have an other set of reviewers

> 
> * xen/arm: add a primitive FF-A mediator
> 
> I'll try to take this one in steps, all structs in one commit, unfragmented sharing in one and finally fragmented sharing in the last.
> 
> * xen/arm: ffa: support sharing memory

Agree

File: xen/arch/arm/include/asm/smccc.h
Comment (bertrand-marquis) [reply xen/arch/arm/include/asm/smccc.h:281]: agree after looking at the rest it is already done this way.

File: xen/arch/arm/vsmc.c
Comment (bertrand-marquis) [reply xen/arch/arm/vsmc.c:89]: nobody commented on that part and i do not think we should claim to support 1.2 if (or until) all calls are not following the 1.2 standard otherwise a guest might have wrong expectations.

**Comment by jenswikl** [comment]
Thanks

File: xen/arch/arm/vsmc.c
Comment (jenswikl) [reply xen/arch/arm/vsmc.c:89]: This is needed by the Linux kernel FF-A driver.
As far as I understand we are compliant with SMCCC version 1.2. If you see anything mandatory missing please let me know and I'll add it.

Comment (bertrand-marquis) [reply xen/arch/arm/vsmc.c:89]: 1.2 is saying that the non used registers should be preserved so nothing needs to be changed unless we have calls using more registers. So it is ok.

Comment (jenswikl) [reply xen/arch/arm/vsmc.c:89]: Thanks

File: xen/arch/arm/include/asm/smccc.h
Comment (bertrand-marquis) [inline xen/arch/arm/include/asm/smccc.h]: Coding style, use /*

File: xen/arch/arm/include/asm/smccc.h
Comment (bertrand-marquis) [inline xen/arch/arm/include/asm/smccc.h]: Coding style use /*

**Review by bertrand-marquis — CHANGES_REQUESTED** [review]
Just some minor coding style

File: xen/arch/arm/Kconfig
Comment (bertrand-marquis) [inline xen/arch/arm/Kconfig]: Please make this unsupported for now, the limitation you state on the possibility of a CPU being held for to long is a good justification for that

File: xen/arch/arm/include/asm/ffa.h
Comment (bertrand-marquis) [inline xen/arch/arm/include/asm/ffa.h]: Wrong file name

File: xen/arch/arm/include/asm/ffa.h
Comment (bertrand-marquis) [inline xen/arch/arm/include/asm/ffa.h]: Please use SPDX style instead of having the full license text in the file

File: xen/arch/arm/include/asm/ffa.h
Comment (bertrand-marquis) [inline xen/arch/arm/include/asm/ffa.h]: If FFA is not activated, init will always fail here.
This should return 0

File: xen/arch/arm/vsmc.c
Comment (bertrand-marquis) [inline xen/arch/arm/vsmc.c]: You are calling a function in a constant, this is really not expected

File: xen/arch/arm/vsmc.c
Comment (bertrand-marquis) [inline xen/arch/arm/vsmc.c]: Please create a proper define for 0x1fU

File: xen/include/public/arch-arm.h
Comment (bertrand-marquis) [inline xen/include/public/arch-arm.h]: You are introducing a generic value "flags" in a public header.
I am wondering if it would not make more sense to add a:
#define XEN_DOMCTL_CONFIG_TEE_FFA 2
At the end having optee support and FFA support in the same guest does not make sense

File: xen/arch/arm/ffa.c
Comment (bertrand-marquis) [inline xen/arch/arm/ffa.c]: Please use SPDX instead of the whole license text

File: xen/arch/arm/ffa.c
Comment (bertrand-marquis) [inline xen/arch/arm/ffa.c]: Why do you need a specific page size here instead of using PAGE_SIZE ?

File: xen/arch/arm/ffa.c
Comment (bertrand-marquis) [inline xen/arch/arm/ffa.c]: Shouldn't the value here depend on page size ? something like 512K / page_size + 1 ?

File: xen/arch/arm/ffa.c
Comment (bertrand-marquis) [inline xen/arch/arm/ffa.c]: We are defining an arbitrary value which i suppose come from the spec, we might want to reference where in the spec we can go to check this, same for the RW after

Comment (bertrand-marquis) [reply xen/arch/arm/ffa.c]: in fact all values here after would need to be checked against the spec, the commit message should clearly reference the spec ID and version used

File: xen/arch/arm/ffa.c
Comment (bertrand-marquis) [inline xen/arch/arm/ffa.c]: domain_id is of type domid_t which is uint16_t so you might have an overflow

File: xen/arch/arm/ffa.c
Comment (bertrand-marquis) [inline xen/arch/arm/ffa.c]: any specific reason to use u16 instead of uint16_t here ?

File: xen/arch/arm/ffa.c
Comment (bertrand-marquis) [inline xen/arch/arm/ffa.c]: This must be converted into a proper define instead of having a function for this.
It will solve the comment made on the header

File: xen/arch/arm/ffa.c
Comment (bertrand-marquis) [inline xen/arch/arm/ffa.c]: If you do this you could just use PAGE_SIZE instead in the rest of the code.
BUILD_BUG_ON(PAGE_SIZE != 4k) adapting the comment

File: xen/arch/arm/ffa.c
Comment (bertrand-marquis) [inline xen/arch/arm/ffa.c]: Please try to introduce the defines when you introduce the support for the function requiring it

**Review by bertrand-marquis — CHANGES_REQUESTED** [review]
In the commit message you cannot reference "latest", you must give the version you used instead to make sure that someone coming in the future would look for the same document.

File: xen/arch/arm/ffa.c
Comment (jenswikl) [reply xen/arch/arm/ffa.c]: Because FF-A always uses 4k page size while Xen may use some other size depending on the configuration.

Comment (bertrand-marquis) [reply xen/arch/arm/ffa.c]: I understand that but you have a build_bug_on if the page_size is not 4k.
Do you expect that we could one day support a Xen page size not 4k but FFA using 4k pages ?

Comment (jenswikl) [reply xen/arch/arm/ffa.c]: Yes

File: xen/arch/arm/ffa.c
Comment (bertrand-marquis) [reply xen/arch/arm/ffa.c]: Please fix the comment here then, it is not FFA_PAGE_SIZE that is defined but FFA supporting only 4k page sizes and the code currently does support having different page sizes in xen and ffa

File: xen/arch/arm/ffa.c
Comment (bertrand-marquis) [reply xen/arch/arm/ffa.c]: ok then please add a comment here to say FFA only supports 4k page size and make the comment on the build_bug_on a bit clearer (as i commented down)

File: xen/include/public/arch-arm.h
Comment (jenswikl) [reply xen/include/public/arch-arm.h]: While I agree that it would be nice to avoid adding another field it becomes a bit confusing.
How about leaving this as is and discussing it on the mailing list when posted?

File: xen/arch/arm/ffa.c
Comment (bertrand-marquis) [inline xen/arch/arm/ffa.c]: I am wondering if here we should not in fact set ourself the vm_id
A guest no knowing that it is running as a VM would probably write 0 here.
I will discuss that with other at Arm to check what they think.

File: xen/arch/arm/ffa.c
Comment (bertrand-marquis) [inline xen/arch/arm/ffa.c]: I think here we should put the final value we will support as define instead of increasing this at each patch

File: xen/arch/arm/ffa.c
Comment (bertrand-marquis) [inline xen/arch/arm/ffa.c]: in the arm64 case, don't we expect to have support for REQ_64 ?

**Review by bertrand-marquis — CHANGES_REQUESTED** [review]
You should also mention in the commit message that you add support to check that the features we need from the TEE are actually supported.

File: xen/include/public/arch-arm.h
Comment (bertrand-marquis) [reply xen/include/public/arch-arm.h]: If you look in the mailing list you will see that there is a serie to add SVE which is ending in the same kind of question.
What we can do is keep it as is for now and wait for the decision on the SVE serie and then do exactly as decided for SVE :-)

File: xen/arch/arm/ffa.c
Comment (bertrand-marquis) [reply xen/arch/arm/ffa.c]: in fact a guest should retrieve it using FFA_GET_ID so this is ok.

File: xen/arch/arm/ffa.c
Comment (jenswikl) [reply xen/arch/arm/ffa.c]: Yeah, `FFA_PAGE_SIZE`. Anyway, I'll update it.

File: xen/arch/arm/ffa.c
Comment (jenswikl) [reply xen/arch/arm/ffa.c]: Yeah, it's painful when reworking the patches, but it is disturbing with a mismatch too when reviewing.

File: xen/arch/arm/ffa.c
Comment (jenswikl) [reply xen/arch/arm/ffa.c]: We are not using FFA_MSG_SEND_DIRECT_REQ_64 in the mediator, so things work even if the SP doesn't support FFA_MSG_SEND_DIRECT_REQ_64.

**Comment by jenswikl** [comment]
Addressed review comments and rebased to be more aligned with what will be posted later.

> You should also mention in the commit message that you add support to check that the features we need from the TEE are actually supported.

I missed this comment, I'll take care of that too.

**Review by bertrand-marquis — APPROVED** [review]

File: xen/arch/arm/Kconfig
Comment (bertrand-marquis) [inline xen/arch/arm/Kconfig]: Convention is to have (unsupported) in the name

File: xen/arch/arm/domain.c
Comment (bertrand-marquis) [inline xen/arch/arm/domain.c]: I think the plan was to use a new value in tee_type here instead of introducing flags

File: xen/arch/arm/domain_build.c
Comment (bertrand-marquis) [inline xen/arch/arm/domain_build.c]: same here, please use tee_type

File: xen/arch/arm/include/asm/domain.h
Comment (bertrand-marquis) [inline xen/arch/arm/include/asm/domain.h]: if we use tee_type we might also use tee here

File: xen/include/public/arch-arm.h
Comment (bertrand-marquis) [reply xen/include/public/arch-arm.h]: as stated before, we should definitely use tee_type instead and introduce a new value for it

**Review by bertrand-marquis — CHANGES_REQUESTED** [review]
Please check to use tee_type instead of introducing flags

**Comment by jenswikl** [comment]
Updated to use tee_type instead.

**Comment by jenswikl** [comment]
Fixed the comment about "(UNSUPPORTED)". I believe I've addressed all comments now.

File: xen/arch/arm/domain.c
Comment (bertrand-marquis) [inline xen/arch/arm/domain.c]: You should have a look into arch/arm/tee/tee.c as it would be more logical to setup ffa as a tee mediator.
This would also mean to move ffa.c into tee directory an dwould prevent most of the changes in the arm common code.
The only change that i can see is the fact that you will need to add a tee_domain_destroy but that would make sense as there is already a tee_domain_init

Comment (jenswikl) [reply xen/arch/arm/domain.c]: `ffa_handle_call()` and `tee_handle_call()` use different SMC ID ranges, so we'd still need the changes in `handle_sssc()`. Is that OK?

Comment (bertrand-marquis) [reply xen/arch/arm/domain.c]: Yes i think there is not other choice for that part. I currently have no idea how to make that part of the code cleaner.
Maybe during upstream review someone will have a better idea.

**Comment by jenswikl** [comment]
Updated to use the TEE mediator framework.

File: xen/arch/arm/vsmc.c
Comment (bertrand-marquis) [inline xen/arch/arm/vsmc.c]: to keep the alphabetical order this should be just before asm/tee/tee.h

**Review by bertrand-marquis — APPROVED** [review]
Just one minor comment with can be fixed before pushing to mailing list (for patch 1)

**Comment by jenswikl** [comment]
Addressing the last comment.

File: xen/arch/arm/tee/ffa.c
Comment (bertrand-marquis) [inline xen/arch/arm/tee/ffa.c:1604]: This will prevent the guest from being created and we should say why.
I would suggest to return something different than ENODEV (EINVAL ?) and add an error print.
We might need to discuss this with the community, either domain_id 0 could be left unused by Xen or UINT16_MAX could be.

File: xen/arch/arm/tee/ffa.c
Comment (bertrand-marquis) [inline xen/arch/arm/tee/ffa.c:885]: Here we are generating a potential infinite loop. why do we need to report back ?

**Review by bertrand-marquis — COMMENTED** [review]
Reviewed until direct request support.
Will try to continue tomorrow but looks good :-)

**Comment by jenswikl** [comment]
I'm sorry I thought we were done for now. I've posted this as v7.

**Comment by bertrand-marquis** [comment]
> I'm sorry I thought we were done for now. I've posted this as v7.

I think this is ok, we can switch to do the review upstream now.

**Pull request closed without merging**
